pre_commit/clientlib.py

```diff
@@ -112,6 +112,24 @@ def validate_manifest_main(argv: Optional[Sequence[str]] = None) -> int:
 META = 'meta'
 
 
+class CheckMutableRev:
+    key = 'rev'
+
+    def check(self, dct: Dict[str, Any]) -> None:
+        if any(
+            mutable in dct.get(self.key, '')
+            for mutable in {'HEAD', 'stable', 'master'}
```

Comment:
I'm not sure this will catch all the cases, notably it's missing `''`, `'main'` at the very least

I wonder if there's an easier way to catch these -- maybe match anything that doesn't look like a version?

Reply:
alternative idea (not added yet):

```python
rev = dct.get(self.key, "")

if (
    "." not in rev and (  # not a version
        len(rev) != 40 or  # full sha
        len(rev) != 7 or  # short sha
        len(rev) != 8 or  # gitlab lists 8 char long sha in the commit-list
    ) or
    not re.match(r"^[a-zA-Z0-9]+$", rev)  # contains other chars (e.g. "_", "-", …)
)
# print warning
```

Reply:
Yeah that might be better -- I wouldn't validate length at all though -- you could have a 15 character "short sha" for example so that would lead to a false positive

It might be worth trying to implement some sort of heuristic and then searching github to find if things adhere to it?

Reply:
so i took a look at current usages of `rev` in `.pre-commit-config.yaml`-files (via github api [`https://api.github.com/search/code?q=filename:%2Epre-commit-config%2Eyaml&page={page}`]) and extracted all revs field. here's what i got so far:

- we can detect shas in general with `r"^[a-z0-9]+$"`, as they all only contain **lowercase** letters and numbers. this will result in false positives for branches, tho.
- version tags are all messed up. most of them follow the pattern `r"^v?([0-9]+\.)+[0-9]$"`, but:
  - some (e.g. ["20.8b1"](https://github.com/psf/black/releases/tag/20.8b1)) contain letters
  - some (e.g. ["4.3.21-2"](https://github.com/PyCQA/isort/releases/tag/4.3.21-2)) contain dashes
  thus, the only reasonable way to detect tags - in my opinion - is, to whitelist all strings that contain dots. (e.g contains `r"[^\.]\.[^\.]"`)
- negative examples i found are:
  - `"stable"`
  - `"master"`
  - `""`

thus, this is my suggestion (tested it against the data i got from the api calls):
```python
bad_revs = {
    "stable",
    "master",
    "HEAD",
}
rev = dct.get(self.key, "")

if "." not in rev and (
    not re.match(r"^[a-z0-9]+$", rev)
    or any(bad_rev in rev for bad_rev in bad_revs)
    or not rev
):
    print(f"Warning: {rev}")
```